src/builder.rs

```diff
@@ -347,6 +347,60 @@ impl<'ctx> Builder<'ctx> {
         PointerValue::new(value)
     }
 
+    /// Alignment arguments are specified in bytes, and should always be a power of 2.
```

Comment:
Please be sure to test what happens when you provide a non-power of two. It may cause the function to return a null ptr (which needs to be handled) or worse cause UB - in which case we need to validate the input and return an `Err` in such cases. (Rust ints have an `is_power_of_two` method)

Reply:
I tried this out, and the module fails verification with:

```
alignment of arg 0 of memory intrinsic must be 0 or a power of 2
  call void @llvm.memcpy.p0i8.p0i8.i64(i8* align 3 %1, i8* align 3 %2, i64 8, i1 false)
```

Given that LLVM already has a check for this with a descriptive error message, do we need a separate test for it?

Reply:
Have you also tried really large values of alignment? Looks like LLVM might assert: https://github.com/llvm/llvm-project/blob/70c0a9675dcaeb74a114a129f3d998bd6dee926f/llvm/lib/IR/Attributes.cpp#L146-L149 which would be really bad on the Rust end.

Reply:
* 0 and 1 both work
* 3 fails validation
* `std::u32::MAX` fails validation

This is surprising to me, considering it looks like it's [hardcoded in a .h file](https://github.com/llvm/llvm-project/blob/f16f139db40e6bf6462ca831eb1ec423c50aeef2/llvm/include/llvm/IR/Value.h#L683) to be [equal to 536870912](https://github.com/llvm/llvm-project/blob/0e62011df891d0e7ad904524edf705d07d12d5d4/llvm/unittests/IR/ValueTest.cpp#L63) and yet it accepts both 536870912 and 2 * 536870912 without complaining.

At any rate, I haven't found an alignment value where it does anything other than either accepting it or failing validation.

Reply:
Really weird. I'm gonna give it a test myself too just to be sure. In the meantime, the tests need to be updated to only run on llvm8

Reply:
You should also write tests that use those sample values, ie `0`, `1`, `3`, `std::u32::max_value()`. However, you don't need to make them as in-depth as the current test with an execution engine. Just create a really simple function that uses your added functions with invalid values and assert that the module fails validation because of it. That way we're more likely to catch UB - if any - at some point.

Reply:
Added in 1ac9cfdc7917191f721e267f59c2bb9e46e52f41